ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java

```diff
@@ -0,0 +1,269 @@
+    URI basePath = new Path(baseDir).toUri();
+    String dataPath = currentPath.replaceFirst(Path.SEPARATOR, basePath.getPath() + Path.SEPARATOR);
+    Path dataLocation = new Path(basePath.getScheme(), basePath.getAuthority(), dataPath);
+    LOG.debug("incoming location: {} , new location: {}", location, dataLocation.toString());
+    return dataLocation.toString();
+  }
+
+  public static class Writer implements Closeable {
+    private static Logger LOG = LoggerFactory.getLogger(Writer.class);
+    private final HiveConf hiveConf;
+    private final Path writePath;
+    private final Boolean excludeExternalTables, dumpMetadataOnly;
+    private OutputStream writer;
+
+    Writer(Path dbRoot, HiveConf hiveConf) throws IOException {
+      this.hiveConf = hiveConf;
+      writePath = new Path(dbRoot, FILE_NAME);
+      excludeExternalTables = !hiveConf.getBoolVar(HiveConf.ConfVars.REPL_INCLUDE_EXTERNAL_TABLES);
+      dumpMetadataOnly = hiveConf.getBoolVar(HiveConf.ConfVars.REPL_DUMP_METADATA_ONLY);
+      if (shouldWrite()) {
+        this.writer = FileSystem.get(hiveConf).create(writePath);
+      }
+    }
+
+    private boolean shouldWrite() {
+      return !dumpMetadataOnly && !excludeExternalTables;
+    }
+
+    /**
+     * this will dump a single line per external table. it can include additional lines for the same
+     * table if the table is partitioned and the partition location is outside the table.
+     */
+    void dataLocationDump(Table table)
+        throws InterruptedException, IOException, HiveException {
+      if (!shouldWrite()) {
+        return;
+      }
+      if (!TableType.EXTERNAL_TABLE.equals(table.getTableType())) {
+        throw new IllegalArgumentException(
+            "only External tables can be writen via this writer, provided table is " + table
+                .getTableType());
+      }
+      Path fullyQualifiedDataLocation =
+          PathBuilder.fullyQualifiedHDFSUri(table.getDataLocation(), FileSystem.get(hiveConf));
+      write(lineFor(table.getTableName(), fullyQualifiedDataLocation, hiveConf));
+      if (table.isPartitioned()) {
+        List<Partition> partitions = Hive.get(hiveConf).getPartitions(table);
+        for (Partition partition : partitions) {
+          boolean partitionLocOutsideTableLoc = !FileUtils
+              .isPathWithinSubtree(partition.getDataLocation(), table.getDataLocation());
+          if (partitionLocOutsideTableLoc) {
+            fullyQualifiedDataLocation = PathBuilder
+                .fullyQualifiedHDFSUri(partition.getDataLocation(), FileSystem.get(hiveConf));
+            write(lineFor(table.getTableName(), fullyQualifiedDataLocation, hiveConf));
+          }
+        }
+      }
+    }
+
+    private static String lineFor(String tableName, Path dataLoc, HiveConf hiveConf)
+        throws IOException, SemanticException {
+      StringWriter lineToWrite = new StringWriter();
+      lineToWrite.append(tableName).append(FIELD_SEPARATOR);
```

Comment:
both table + partition locations work with the same logic as above. its for debuggability to identify what path is actually associated for a table that might cause problem. you see issues with including this ?

Reply:
No, I don't have issues including this information even if it's not used. It's good for debugging. But if we include partition name as well along with the table name when we are dumping location of a partition, that will further improve readability. That also avoids any confusion in case somebody want to use this information in future. Right now, having two paths associated with a table is confusing esp. when both the paths point to different locations (e.g. partition not within the table directory).

Reply:
yeh i was thinking we could use the location to get the partition, but as you said if location is explicitly specified then find that partition in a large table will be troublesome. we should try to incorporate this in the next task of external tables, since this is not breaking any functionality for now.